### PR TITLE
Better docs

### PR DESCRIPTION
**Pull request by jowezarek**
This PR removes >50 hard-coded `.rst` files containing manually added method and class names.

Instead, we make use of the `:toctree:` directive of the `autosummary` extension to generate `.rst` files for each Psydac module.

Changing the corresponding `module.rst` template of the `autosummary` extension, we are able to include the `.. inheritance-diagram::` directive to the generated files and also slighty modify their structure and appearance.

Finally, we modify the `__init__` of `AutosummaryRenderer` to include a filter that allows us to reduce `psydac.module.submodule` entries in the navigation bar to `module.submodule` entries.

For @pyccel/psydac-dev reviewers: To see the "new" documentation, scroll down to `checks` and click on `Details` of the `documentation / build_docs (pull_request) ` check. Then, click on `Summary`. Scroll down and download `github-pages` in the `Artifacts` tab. Unpack `github-pages.zip`, in `github-pages` unpack `artifact.tar`, and finally in `artifact` open `index.html`.

### CODE REVIEW

File: docs/source/modules/api.rst [hunk 1]
```diff
@@ -1,41 +1,18 @@
...
+    psydac.api.basic
+    psydac.api.discretization
+    psydac.api.equation
+    psydac.api.essential_bc
+    psydac.api.expr
+    psydac.api.feec
+    psydac.api.fem
+    psydac.api.glt
+    psydac.api.grid
+    psydac.api.postprocessing
```
Comment (yguclu) [inline docs/source/modules/api.rst:18]: Have you tried to use the `currentmodule` directive?
```rst
.. currentmodule: psydac.api
.. autosummary::
    :nosignatures:
    :toctree: STUBDIR
    :template: autosummary/module.rst

    basic
    discretization
    equation
    essential_bc
    expr
    feec
    fem
    glt
    grid
    postprocessing
````

Comment (jowezarek) [reply docs/source/modules/api.rst:18]: That's a really good point, I had forgotten about the `.. currentmodule::` directive!
I propose writing 
```
.. currentmodule: psydac
.. autosummary::
    :nosignatures:
    :toctree: STUBDIR
    :template: autosummary/module.rst

    api.basic
    api.discretization
    api.equation
    api.essential_bc
    api.expr
    api.feec
    api.fem
    api.glt
    api.grid
    api.postprocessing
```

As we have several modules named `basic` for example. This way, they will appear everywhere as `submodule.basic`, both in the Navigation sidebar and on the module overview page.

**Comment by FrederikSchnack** [comment]
Looks good! However I couldn't find the submodules like psydac.feec.multipatch is this on purpose?

**Comment by jowezarek** [comment]
> Looks good! However I couldn't find submodules like psydac.feec.multipatch is this on purpose?

Thanks, that is something I wanted to talk with @yguclu about but forgot!
I have so far manually excluded
- `psydac.cad`, `psydac.cmd`, `psydac.core`, `psydac.pyccel` (and `psydac.pyccel`'s submodules)
- `psydac.api.ast`, `psydac.api.printing`, `psydac.feec.multipatch`, `psydac.linalg.kernels`

@yguclu which of these should we include in the documentation?

**Comment by yguclu** [comment]
> > Looks good! However I couldn't find submodules like psydac.feec.multipatch is this on purpose?
> 
> Thanks, that is something I wanted to talk with @yguclu about but forgot! I have so far manually excluded
> 
>     * `psydac.cad`, `psydac.cmd`, `psydac.core`, `psydac.pyccel` (and `psydac.pyccel`'s submodules)
> 
>     * `psydac.api.ast`, `psydac.api.printing`, `psydac.feec.multipatch`, `psydac.linalg.kernels`
> 
> 
> @yguclu which of these should we include in the documentation?

For now I would only exclude `psydac.pyccel` because that contains a lot of messy code. 

`psydac.api.ast` and `psydac.api.printing` constitute the code generation part of Psydac, which will be useful to anyone who wants to improve our assembly algorithms, for example...

**Comment by vcarlier** [comment]
Looks good! Thanks Julian!

I only have a small comment, when I go to a submodule by clicking on the link on the parent module, I don't see the overall description of the module but directly the doc for the first class of the module.

For example, if I open the documentation, click on `modules` at the bottom right of the page, then click on `api` and on `api.basic` (at the center of the page, not on the index), I go to the documentation of `BasicCodeGen` instead of seeing the general documentation of the module and the class diagram.

I think this is because the link goes to `file:///home/vcarlier/Downloads/modules/STUBDIR/psydac.api.basic.html#module-0` when it should probably go to `file:///home/vcarlier/Downloads/modules/STUBDIR/psydac.api.basic.htm` ?

**Review by campospinto — COMMENTED** [review]
Thanks Julian! This is very useful.

**Comment by jowezarek** [comment]
> Looks good! Thanks Julian!
> 
> I only have a small comment, when I go to a submodule by clicking on the link on the parent module, I don't see the overall description of the module but directly the doc for the first class of the module.
> 
> For example, if I open the documentation, click on `modules` at the bottom right of the page, then click on `api` and on `api.basic` (at the center of the page, not on the index), I go to the documentation of `BasicCodeGen` instead of seeing the general documentation of the module and the class diagram.
> 
> I think this is because the link goes to `file:///home/vcarlier/Downloads/modules/STUBDIR/psydac.api.basic.html#module-0` when it should probably go to `file:///home/vcarlier/Downloads/modules/STUBDIR/psydac.api.basic.htm` ?

I've noticed that, too, but unfortunately don't know how to fix it. When clicking on `api.basic` for example, the link points to a particular instance of `.. automodule:: psydac.api.basic`. We designed the documentation in such a way that this instance appears last on the page (actually that last part, everything under "Details", makes up most of the page, but it's still technically the last content block). So unless we change the order, I don't think I can fix this :/

**Comment by jowezarek** [comment]
> > > Looks good! However I couldn't find submodules like psydac.feec.multipatch is this on purpose?
> > 
> > 
> > Thanks, that is something I wanted to talk with @yguclu about but forgot! I have so far manually excluded
> > ```
> > * `psydac.cad`, `psydac.cmd`, `psydac.core`, `psydac.pyccel` (and `psydac.pyccel`'s submodules)
> > 
> > * `psydac.api.ast`, `psydac.api.printing`, `psydac.feec.multipatch`, `psydac.linalg.kernels`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @yguclu which of these should we include in the documentation?
> 
> For now I would only exclude `psydac.pyccel` because that contains a lot of messy code.
> 
> `psydac.api.ast` and `psydac.api.printing` constitute the code generation part of Psydac, which will be useful to anyone who wants to improve our assembly algorithms, for example...

I've included (almost) successfully `api.ast`, `api.printing`, `feec.multipatch` and `linalg.kernels` in the documentation when I noticed that for some reason `feec.multipatch.multipatch_domain_utilities` and `feec.multipatch.plotting_utilities` where not properly documented.

This eventually caused me to "fix" all of our `__init__` files, for example I changed `psydac.feec.__init__.py` from 
```
__all__ = ['derivatives']

from psydac.feec import derivatives
```
to
```
__all__ = ('derivatives', 'global_projectors', 'multipatch', 'pull_push', 'pushforward')

from psydac.feec import derivatives
from psydac.feec import global_projectors
from psydac.feec import multipatch
from psydac.feec import pull_push
from psydac.feec import pushforward
```
I did so with all instances of `__init__.py`. This however caused sphinx to not find `psydac.api` anymore. Eventually, I figured that this is due to a circular import error:
```
>>> import psydac.api
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jubuntu/psydac/psydac/__init__.py", line 6, in <module>
    from psydac import api
  File "/home/jubuntu/psydac/psydac/api/__init__.py", line 6, in <module>
    from psydac.api import discretization
  File "/home/jubuntu/psydac/psydac/api/discretization.py", line 32, in <module>
    from psydac.api.feec         import DiscreteDerham
  File "/home/jubuntu/psydac/psydac/api/feec.py", line 4, in <module>
    from psydac.feec.derivatives       import Derivative_1D, Gradient_2D, Gradient_3D
  File "/home/jubuntu/psydac/psydac/feec/__init__.py", line 5, in <module>
    from psydac.feec import multipatch
  File "/home/jubuntu/psydac/psydac/feec/multipatch/__init__.py", line 4, in <module>
    from psydac.feec.multipatch import api
  File "/home/jubuntu/psydac/psydac/feec/multipatch/api.py", line 14, in <module>
    from psydac.api.discretization        import discretize as discretize_single_patch
ImportError: cannot import name 'discretize' from partially initialized module 'psydac.api.discretization' (most likely due to a circular import) (/home/jubuntu/psydac/psydac/api/discretization.py)
```

Any suggestion on what to do about it @yguclu?

**Comment by e-moral-sanchez** [comment]
Looks very cool! I couldn't find anything strange

**Comment by yguclu** [comment]
> > Looks good! Thanks Julian!
> > I only have a small comment, when I go to a submodule by clicking on the link on the parent module, I don't see the overall description of the module but directly the doc for the first class of the module.
> > For example, if I open the documentation, click on `modules` at the bottom right of the page, then click on `api` and on `api.basic` (at the center of the page, not on the index), I go to the documentation of `BasicCodeGen` instead of seeing the general documentation of the module and the class diagram.
> > I think this is because the link goes to `file:///home/vcarlier/Downloads/modules/STUBDIR/psydac.api.basic.html#module-0` when it should probably go to `file:///home/vcarlier/Downloads/modules/STUBDIR/psydac.api.basic.html` ?
> 
> I've noticed that, too, but unfortunately don't know how to fix it. When clicking on `api.basic` for example, the link points to a particular instance of `.. automodule:: psydac.api.basic`. We designed the documentation in such a way that this instance appears last on the page (actually that last part, everything under "Details", makes up most of the page, but it's still technically the last content block). So unless we change the order, I don't think I can fix this :/

@jowezarek This might be a small problem, but I find it quite distracting when navigating the documentation. We are using an `.. automodule` command at the top of the file, so it should be possible to get a link to `[...]/STUBDIR/psydac.api.basic.html#module-psydac.api.basic` at least?

**Comment by ratnania** [comment]
Thanks @jowezarek  for this great and nice work! we really need it!
BTW, just a naive question; why rst+sphinx and not markdown+mkdocs? 
I know that originally our documentation was in rst, but since we're already using md in pyccel, it might be better to stick to the same technology? that will make things easier for everybody to contribute ...

**Comment by yguclu** [comment]
> Thanks @jowezarek for this great and nice work! we really need it! BTW, just a naive question; why rst+sphinx and not markdown+mkdocs? I know that originally our documentation was in rst, but since we're already using md in pyccel, it might be better to stick to the same technology? that will make things easier for everybody to contribute ...

The answer is simple, rst is used across many projects at NMPP. In particular, Sphinx is used by the Struphy project, so we are leveraging their expertise with this tool. Our main goal is to bring our documentation to the same level as the other codes at NMPP, with minimal user intervention (we are focusing on automatic documentation). In the future we could switch to a different technology, if this can provide the same functionality (i.e. if we can produce the same content within the html files).

As for contributing, everyone should focus on docstrings now. That is independent of the documentation build tool.

**Comment by EmilyBourne** [comment]
> Thanks @jowezarek  for this great and nice work! we really need it!
> BTW, just a naive question; why rst+sphinx and not markdown+mkdocs? 
> I know that originally our documentation was in rst, but since we're already using md in pyccel, it might be better to stick to the same technology? that will make things easier for everybody to contribute ...

Just to add some context from the Pyccel side, we are using markdown because it is handled by GitHub. This is important for us precisely because we don't have any clean html generated docs. When we finally add some we will probably also use sphinx as there are plugins that can parse our docstrings and plugins to use markdown

**Comment by jowezarek** [comment]
> > > Looks good! Thanks Julian!
> > > I only have a small comment, when I go to a submodule by clicking on the link on the parent module, I don't see the overall description of the module but directly the doc for the first class of the module.
> > > For example, if I open the documentation, click on `modules` at the bottom right of the page, then click on `api` and on `api.basic` (at the center of the page, not on the index), I go to the documentation of `BasicCodeGen` instead of seeing the general documentation of the module and the class diagram.
> > > I think this is because the link goes to `file:///home/vcarlier/Downloads/modules/STUBDIR/psydac.api.basic.html#module-0` when it should probably go to `file:///home/vcarlier/Downloads/modules/STUBDIR/psydac.api.basic.html` ?
> > 
> > 
> > I've noticed that, too, but unfortunately don't know how to fix it. When clicking on `api.basic` for example, the link points to a particular instance of `.. automodule:: psydac.api.basic`. We designed the documentation in such a way that this instance appears last on the page (actually that last part, everything under "Details", makes up most of the page, but it's still technically the last content block). So unless we change the order, I don't think I can fix this :/
> 
> @jowezarek This might be a small problem, but I find it quite distracting when navigating the documentation. We are using an `.. automodule` command at the top of the file, so it should be possible to get a link to `[...]/STUBDIR/psydac.api.basic.html#module-psydac.api.basic` at least?
> 
> Alternatively, we could write a simple Python script which navigates through the generated `.rst` files and removes the string `#module-0` from those links.

So I had a look again and am still very uncertain whether we can achieve that.

Adding the `:no-index:` directive to either of the two instances of `.. automodule::` directive gets rid of a bunch of "double-reference"-errors which is nice to start with. 

Adding `:no-index` to the second instance of `.. automodule::` does indeed have as consequence, that the links now point to the top of the module page. However, all links to documented classes, methods, attributes, ... vanish as well, so this is not an option. I added `:no-index:` to the first instance anyways.

`#module-0` was probably an artifact of not having used `:no-index:` at all.

What would need to be done, is change `href="STUBDIR/psydac.[module].[submodule].html#module-psydac.linalg.basic"` to `href="STUBDIR/psydac.[module].[submodule].html"` several times in several files in `docs/build/html/modules`.

But I don't know how to change files in `/build`, as it does not appear on our github.

**Review by yguclu — APPROVED** [review]
Great job!

**Pull request merged**